packages/definitions-parser/src/lib/module-info.ts

```diff
@@ -130,149 +130,141 @@ export function allReferencedFiles(
   entryFilenames: readonly string[],
   fs: FS,
   packageName: string,
-  baseDirectory: string
+  moduleResolutionHost: ts.ModuleResolutionHost,
+  compilerOptions: ts.CompilerOptions,
 ): { types: Map<string, ts.SourceFile>; tests: Map<string, ts.SourceFile>; hasNonRelativeImports: boolean } {
   const seenReferences = new Set<string>();
   const types = new Map<string, ts.SourceFile>();
   const tests = new Map<string, ts.SourceFile>();
+  // The directory where the tsconfig/index.d.ts is - i.e., may be a version within the package
+  const baseDirectory = path.resolve("/", fs.debugPath());
+  // The root of the package and all versions, i.e., the direct subdirectory of types/
+  const packageDirectory = baseDirectory.slice(0, baseDirectory.lastIndexOf(`types/${getMangledNameForScopedPackage(packageName)}`) + `types/${getMangledNameForScopedPackage(packageName)}`.length);
   let hasNonRelativeImports = false;
-  entryFilenames.forEach((text) => recur({ text, exact: true }));
+  entryFilenames.forEach((fileName) => recur(undefined, { text: fileName, kind: "path" }));
   return { types, tests, hasNonRelativeImports };
 
-  function recur({ text, exact }: Reference): void {
-    const resolvedFilename = exact ? text : resolveModule(text, fs);
-    if (seenReferences.has(resolvedFilename)) {
+  function recur(containingFileName: string | undefined, { text, kind, resolutionMode }: Reference): void {
+    // An absolute file name for use with TS resolution, e.g. '/DefinitelyTyped/types/foo/index.d.ts'
+    let resolvedFileName: string | undefined;
+    switch (kind) {
+      case "path":
+        if (containingFileName) {
+          resolvedFileName = ts.resolveTripleslashReference(text, containingFileName);
+        } else {
+          resolvedFileName = path.resolve(baseDirectory, text);
+        }
+        break;
+      case "types":
+        resolvedFileName = ts.resolveTypeReferenceDirective(
+          text,
+          assertDefined(containingFileName, "Must have a containing file to resolve a type reference directive"),
+          compilerOptions,
+          moduleResolutionHost,
+          /*redirectedReference*/ undefined,
+          /*cache*/ undefined,
+          resolutionMode).resolvedTypeReferenceDirective?.resolvedFileName;
+        break;
+      case "import":
+        resolvedFileName = ts.resolveModuleName(
+          text,
+          assertDefined(containingFileName, "Must have an containing file to resolve an import"),
+          compilerOptions,
+          moduleResolutionHost,
+          /*cache*/ undefined,
+          /*redirectedReference*/ undefined,
```

Comment:
hmm, cache and redirectedReference swap positions between resolveTypeReferenceDirective/resolveModuleName

Reply:
Sadly, this is correct